auth/user_mgt.go

```diff
@@ -372,6 +373,82 @@ func processClaims(p map[string]interface{}) error {
 	return nil
 }
 
+// Error handlers.
+
+const (
+	emailAlredyExists        = "email-already-exists"
+	idTokenRevoked           = "id-token-revoked"
+	insufficientPermission   = "insufficient-permission"
+	phoneNumberAlreadyExists = "phone-number-already-exists"
+	projectNotFound          = "project-not-found"
+	uidAlreadyExists         = "uid-already-exists"
+	unknown                  = "unknown-error"
+	userNotFound             = "user-not-found"
+)
+
+// IsEmailAlreadyExists checks if the given error was due to a duplicate email.
+func IsEmailAlreadyExists(err error) bool {
+	return internal.HasErrorCode(err, emailAlredyExists)
+}
+
+// IsIDTokenRevoked checks if the given error was due to a revoked ID token.
+func IsIDTokenRevoked(err error) bool {
+	return internal.HasErrorCode(err, idTokenRevoked)
+}
+
+// IsInsufficientPermission checks if the given error was due to insufficient permissions.
+func IsInsufficientPermission(err error) bool {
+	return internal.HasErrorCode(err, insufficientPermission)
+}
+
+// IsPhoneNumberAlreadyExists checks if the given error was due to a duplicate phone number.
+func IsPhoneNumberAlreadyExists(err error) bool {
+	return internal.HasErrorCode(err, phoneNumberAlreadyExists)
+}
+
+// IsProjectNotFound checks if the given error was due to a non-existing project.
+func IsProjectNotFound(err error) bool {
+	return internal.HasErrorCode(err, projectNotFound)
+}
+
+// IsUIDAlreadyExists checks if the given error was due to a duplicate uid.
+func IsUIDAlreadyExists(err error) bool {
+	return internal.HasErrorCode(err, uidAlreadyExists)
+}
+
+// IsUnknown checks if the given error was due to a unknown server error.
+func IsUnknown(err error) bool {
+	return internal.HasErrorCode(err, unknown)
+}
+
+// IsUserNotFound checks if the given error was due to non-existing user.
+func IsUserNotFound(err error) bool {
+	return internal.HasErrorCode(err, userNotFound)
+}
+
+var serverError = map[string]string{
+	"CONFIGURATION_NOT_FOUND": projectNotFound,
+	"DUPLICATE_EMAIL":         emailAlredyExists,
+	"DUPLICATE_LOCAL_ID":      uidAlreadyExists,
+	"EMAIL_EXISTS":            emailAlredyExists,
+	"INSUFFICIENT_PERMISSION": insufficientPermission,
+	"PHONE_NUMBER_EXISTS":     phoneNumberAlreadyExists,
+	"PROJECT_NOT_FOUND":       projectNotFound,
+}
+
+func handleServerError(err error) *internal.FirebaseError {
+	var serverCode string
+	gerr, ok := err.(*googleapi.Error)
+	if ok {
+		serverCode = gerr.Message
+	}
+	clientCode, ok := serverError[serverCode]
+	if !ok {
+		clientCode = unknown
+	}
+	return internal.Error(clientCode, err.Error())
```

Comment:
It seems like this might get called in cases where the error is generated on the client by a failure to connect or something. Do we want to always wrap error in our own type like this, possibly losing original error data other than the string? Or should we just pass on errors that aren't from the backend as-is? This is something I've been thinking about a lot on all of our SDKs.

Reply:
Good catch. We shouldn't be wrapping non-backend errors. Updated the implementation.